packages/workbox-webpack-plugin/src/index.js

```diff
@@ -59,24 +59,68 @@ class WorkboxWebpackPlugin {
   }
 
   /**
-   * @return {Object} All workbox configuration options that can be accepted
-   * by {@link module:workbox-build.generateSWString}
-   *
+   * @param {Object} compilation The webpack compilation.
    * @private
    */
-  get generateSWStringOptions() {
+  async handleEmit(compilation) {
     const {
-      importScripts = [],
-      // TODO: decide on other options that should be passed to
-      // `generateSWString`. What operations should be possible for webpack
-      // users? eg, should webpack users also be able to specify a
-      // `globDirectory` that allows manifestEntries to be added that are
-      // outside webpack's scope?
+      importWorkboxFrom = 'cdn',
+      swSrc,
     } = this.config;
-
-    return {
-      importScripts,
-    };
+    const serviceWorkerFilename = swSrc ? path.basename(swSrc) : 'sw.js';
+
+    let workboxSWImport;
+    switch (importWorkboxFrom) {
+      case 'cdn': {
+        workboxSWImport = getModuleUrl('workbox-sw');
+        break;
+      }
+
+      case 'local': {
+        // TODO: Implement.
+        break;
+      }
+
+      case 'disabled': {
+        // No-op.
+        break;
+      }
+
+      default: {
+        for (const chunk of compilation.chunks) {
+          if (chunk.name === importWorkboxFrom) {
+            workboxSWImport = chunk.files;
+            this.config.excludeChunks = (this.config.excludeChunks || [])
```

Comment:
could you move this initialisation out of this switch statement just to seperate the initialisation and the addition of the chunk name.